### PR TITLE
New 'build' task to replace current 'html' task

### PR DESCRIPTION
**Pull request by felixzapata**
- update gulp-useref to version 3 (it changes [how to use the assets](https://github.com/jonkemp/gulp-useref))
- new 'build' task to replace 'html' task. This new task runs 'images' and 'fonts' and after that, it searchs for the assets in the html files and optimize them.

This way I think that 'default' task is more clear.

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -155,11 +126,16 @@ gulp.task('fonts', function() {
     }));
 });
 
+
 // Scan your HTML for assets & optimize them
-gulp.task('html', function() {
-  return optimizeHtmlTask(
-    ['app/**/*.html', '!app/{elements,test,bower_components}/**/*.html'],
-    dist());
+gulp.task('build', ['images', 'fonts'], function () {
```
Comment (robdodson) [inline gulpfile.js]: I don't think this task actually depends on `images` or `fonts` having run already does it? In the previous version all 3 tasks were running in parallel, which should be a bit faster if I understand correctly

File: gulpfile.js [hunk 2]
```diff
@@ -67,35 +67,6 @@ var imageOptimizeTask = function(src, dest) {
...
-    .pipe($.if('*.js', $.uglify({
-      preserveComments: 'some'
-    })))
-    // Concatenate and minify styles
-    // In case you are still using useref build blocks
-    .pipe($.if('*.css', $.minifyCss()))
-    .pipe(assets.restore())
-    .pipe($.useref())
-    // Minify any HTML
-    .pipe($.if('*.html', $.minifyHtml({
```
Comment (robdodson) [inline gulpfile.js]: was minifyHtml removed in your PR?

Comment (felixzapata) [inline gulpfile.js]: oopps... yes sorry I removed the minify for the HTML. I will update it..

File: gulpfile.js (hunk 1 above)
Comment (felixzapata) [inline gulpfile.js]: well, in this case is more for to group the tasks related with a "build" task. That is the reason you run 'build' and first to this, 'fonts' and 'images' in parallel.

But you can obtain the same result as the original if rename 'build' to 'html' and mantein in the 'default' the original code.

Comment (robdodson) [inline gulpfile.js]: I think I'm ok with the name change. Honestly I think of the entire gulp file as a 'build' so the term doesn't have much significance for me.. but I like the idea of reducing the number of individual tasks listed in the default file.

Comment (felixzapata) [inline gulpfile.js]: ok, so, it is necessary some other change to the pull request?

**Comment by samccone** [comment]
@felixzapata mind rebasing off of master? (sorry about the delay!)

Once this is rebase I will take it for a spin, if things seem fine we will need to go in and update each of the recipes then we can :shipit:

**Comment by felixzapata** [comment]
done @samccone. I've updated two recipes too, following the new tasks.

**Comment by abdonrd** [comment]
@felixzapata please, rebase on master (now updated with the actual gulpfile.js) and I personally will take a look. :)

**Comment by felixzapata** [comment]
@abdonrd done!

File: docs/add-es2015-support-babel.md [hunk 3]
```diff
@@ -42,16 +42,16 @@ Make sure the `js` gulp task is triggered by the common build tasks:
  - In the gulp `serve` task, make sure `js` is triggered initially and on HTML and JS files changes:
 
 ```patch
-- gulp.task('serve', ['styles'], function() {
-+ gulp.task('serve', ['styles', 'js'], function () {
+-gulp.task('serve', ['styles'], function () {
```
Comment (abdonrd) [inline docs/add-es2015-support-babel.md]: Please, don't change the spaces. We want to keep the same as actual `gulpfile.js`.

space before `gulp` and space after `function`.

File: docs/add-es2015-support-babel.md [hunk 4]
```diff
@@ -42,16 +42,16 @@ Make sure the `js` gulp task is triggered by the common build tasks:
...
++gulp.task('serve', ['styles', 'js'], function () {
```
Comment (abdonrd) [inline docs/add-es2015-support-babel.md]: space before `gulp`

File: docs/add-es2015-support-babel.md [hunk 5]
```diff
@@ -42,16 +42,16 @@ Make sure the `js` gulp task is triggered by the common build tasks:
...
 
   ...
 
 - gulp.watch(['app/**/*.html', '!app/bower_components/**/*.html'], reload);
 + gulp.watch(['app/**/*.html', '!app/bower_components/**/*.html'], ['js', reload]);
   gulp.watch(['app/styles/**/*.css'], ['styles', reload]);
-- gulp.watch(['app/scripts/**/*.js'], reload);
-+ gulp.watch(['app/scripts/**/*.js'], ['js', reload]);
+- gulp.watch(['app/{scripts,elements}/**/{*.js,*.html}'], ['lint']);
++ gulp.watch(['app/{scripts,elements}/**/{*.js,*.html}'], ['lint', 'js']);
```
Comment (abdonrd) [inline docs/add-es2015-support-babel.md]: The actual `gulpfile.js` hasn't `lint` task.

File: docs/add-es2015-support-babel.md [hunk 6]
```diff
@@ -65,8 +65,8 @@ gulp.task('default', ['clean'], function(cb) {
 
   runSequence(
     ['ensureFiles', 'copy', 'styles'],
-+   'js',
-    ['images', 'fonts', 'html'],
+-   ['build'],
```
Comment (abdonrd) [inline docs/add-es2015-support-babel.md]: Remove `build` is wrong

File: docs/add-es2015-support-babel.md [hunk 7]
```diff
@@ -76,18 +76,19 @@ gulp.task('default', ['clean'], function(cb) {
 
  ```patch
  // Scan your HTML for assets & optimize them
- gulp.task('html', function() {
+ gulp.task('html', function () {
```
Comment (abdonrd) [inline docs/add-es2015-support-babel.md]: Space after 'function`.

File: docs/add-es2015-support-babel.md [hunk 8]
```diff
@@ -76,18 +76,19 @@ gulp.task('default', ['clean'], function(cb) {
...
      dist());
  });
  ```
 
+
  - In the `optimizeHtmlTask` function remove the `searchPath` attribute since all assets should be found under the `dist` folder and we want to make sure we are not picking up duplicates and un-transpiled JS files:
 
 ```patch
-var optimizeHtmlTask = function(src, dest) {
+var optimizeHtmlTask = function (src, dest) {
```
Comment (abdonrd) [inline docs/add-es2015-support-babel.md]: Space after 'function`.

File: docs/add-eslint-support.md [hunk 9]
```diff
@@ -77,8 +77,7 @@ Make sure the `lint` gulp task is triggered by the common build tasks:
 
   gulp.watch(['app/**/*.html', '!app/bower_components/**/*.html'], reload);
   gulp.watch(['app/styles/**/*.css'], ['styles', reload]);
-- gulp.watch(['app/scripts/**/*.js'], reload);
-+ gulp.watch(['app/scripts/**/*.js'], ['lint', reload]);
++ gulp.watch(['app/{scripts,elements}/**/{*.js,*.html}'], ['lint', reload]);
```
Comment (abdonrd) [inline docs/add-eslint-support.md]: This is unrelated with `build` task. Please, add a different PR for it.

File: docs/add-es2015-support-babel.md [hunk 10]
```diff
@@ -110,21 +111,29 @@ Need to change the vulcanize command to grab the newly translated files.  .tmp h
...
 ```
 
 - Finally update `vulcanize` task to point to `.tmp` directory:
 ```patch
-gulp.task('vulcanize', function() {
-- return gulp.src('app/elements/elements.html')
-+ return gulp.src('.tmp/elements/elements.html')
-    .pipe($.vulcanize({
-      stripComments: true,
+ // Vulcanize granular configuration
```
Comment (abdonrd) [inline docs/add-es2015-support-babel.md]: Why adding more code? You don't update anything here.

File: docs/add-es2015-support-babel.md [hunk 11]
```diff
@@ -110,21 +111,29 @@ Need to change the vulcanize command to grab the newly translated files.  .tmp h
...
-    ['ensureFiles', 'copy', 'styles'],
-    'js',
+ gulp.task('default', ['clean'], function(cb) {
+   // Uncomment 'cache-config' if you are going to use service workers.
+   runSequence(
++    'bowertotmp',
+     ['ensureFiles', 'copy', 'styles'],
+-    ['build'],
++    ['js', 'build'],
+@@ -324,6 +325,12 @@
```
Comment (abdonrd) [inline docs/add-es2015-support-babel.md]: Why adding this code?

**Comment by abdonrd** [comment]
Minor comments. Thanks @felixzapata! :)

File: docs/add-eslint-support.md (hunk 9 above)
Comment (felixzapata) [inline docs/add-eslint-support.md]: do you want to remove the changes in this recipe? I made changes for both of them to add the new `build` task

File: docs/add-es2015-support-babel.md [hunk 12]
```diff
@@ -54,7 +54,6 @@ Make sure the `js` gulp task is triggered by the common build tasks:
 + gulp.watch(['app/scripts/**/*.js'], ['js', reload]);
   gulp.watch(['app/images/**/*'], reload);
 });
-```
```
Comment (abdonrd) [inline docs/add-es2015-support-babel.md]: Ups! Don't remove this, please. :)

File: docs/add-es2015-support-babel.md [hunk 13]
```diff
@@ -42,7 +42,7 @@ Make sure the `js` gulp task is triggered by the common build tasks:
  - In the gulp `serve` task, make sure `js` is triggered initially and on HTML and JS files changes:
 
 ```patch
-- gulp.task('serve', ['styles'], function() {
+- gulp.task('serve', ['styles'], function () {
```
Comment (abdonrd) [inline docs/add-es2015-support-babel.md]: Sorry, the space after `function`. 😅

File: docs/add-es2015-support-babel.md [hunk 14]
```diff
@@ -66,33 +65,25 @@ gulp.task('default', ['clean'], function(cb) {
   runSequence(
     ['ensureFiles', 'copy', 'styles'],
 +   'js',
-    ['images', 'fonts', 'html'],
+-   ['images', 'fonts', 'html'],
+    'build',
```
Comment (abdonrd) [inline docs/add-es2015-support-babel.md]: Please, the `+` before `build`. :)

File: docs/add-eslint-support.md (hunk 9 above)
Comment (abdonrd) [inline docs/add-eslint-support.md]: But here you only add the linting to the elements folder, nothing concerning the `build` task.

File: gulpfile.js [hunk 15]
```diff
@@ -71,40 +71,12 @@ var imageOptimizeTask = function(src, dest) {
...
-      title: 'html'
-    }));
-};
-
 // Compile and automatically prefix stylesheets
 gulp.task('styles', function() {
   return styleTask('styles', ['**/*.css']);
 });
 
+
```
Comment (abdonrd) [inline gulpfile.js]: Remove extra space, please. :)

File: docs/add-es2015-support-babel.md [hunk 16]
```diff
@@ -66,33 +66,24 @@ gulp.task('default', ['clean'], function(cb) {
   runSequence(
     ['ensureFiles', 'copy', 'styles'],
 +   'js',
-    ['images', 'fonts', 'html'],
+-   ['images', 'fonts', 'html'],
++   'build',
```
Comment (abdonrd) [inline docs/add-es2015-support-babel.md]: @felixzapata sorry, my fault! 😅 
We need to remove the `+` of `'build',`, because the `gulpfile.js` will have that task.

File: docs/add-es2015-support-babel.md [hunk 17]
```diff
@@ -66,33 +66,24 @@ gulp.task('default', ['clean'], function(cb) {
...
 ```
 
- - In the `html` task replace `app` in the paths by `dist` since dist should already contain all JS and HTML files now transpiled.
+ - In the `build` task replace `app` in the paths by `dist` since dist should already contain all JS and HTML files now transpiled.
 
  ```patch
  // Scan your HTML for assets & optimize them
- gulp.task('html', function() {
+ gulp.task('build', ['images', 'fonts'], function() {
    return optimizeHtmlTask(
```
Comment (abdonrd) [inline docs/add-es2015-support-babel.md]: In the new `build` task, we don't have that line. :)

File: docs/add-eslint-support.md (hunk 9 above)
Comment (abdonrd) [inline docs/add-eslint-support.md]: Please, remember move it to another PR.

Comment (abdonrd) [inline docs/add-eslint-support.md]: @felixzapata Please, remember move it to another PR.

**Comment by abdonrd** [comment]
Another minor comments. Thanks @felixzapata! :)

File: docs/add-es2015-support-babel.md (hunk 17 above)
Comment (felixzapata) [inline docs/add-es2015-support-babel.md]: I have not that line in my file. Only I see:

 ```patch
 // Scan your HTML for assets & optimize them
 gulp.task('build', ['images', 'fonts'], function() {
   return optimizeHtmlTask(
-    return gulp.src(['app/**/*.html', '!app/{elements,test,bower_components}/**/*.html'])
+    return gulp.src(['dist/**/*.html', '!dist/{elements,test,bower_components}/**/*.html'])

  ...
 ```

Comment (abdonrd) [inline docs/add-es2015-support-babel.md]: When you create the `build` task, you don't have the `return optimizeHtmlTask(` line:

screen shot 2016-05-02 at 11 28 10

File: docs/add-eslint-support.md (hunk 9 above)
Comment (abdonrd) [inline docs/add-eslint-support.md]: @felixzapata Please, remember move it to another PR.

File: docs/add-es2015-support-babel.md (hunk 16 above)
Comment (abdonrd) [inline docs/add-es2015-support-babel.md]: Friendly ping to @felixzapata. :)

**Comment by abdonrd** [comment]
Thanks @felixzapata!

**Comment by gazal-k** [comment]
`app/scripts/app.js` is not getting copied to `dist/scripts` post this change

**Comment by felixzapata** [comment]
hi @gazal-k , do you mean when you run `gulp serve` or `gulp serve:dist`?

**Comment by gazal-k** [comment]
This new `build` task and the earlier `html` task was only getting called from `default` task. I found the problem when running the `default` task, but I'm assuming it would also affect `serve:dist` coz it depends on `default` but not `serve`

**Comment by felixzapata** [comment]
If you run `gulp serve`, the app.js file is served from the `app` folder. 

In the current release version (1.3.0.. with the old task) the `app.js` file is only copied when the `gulp serve:dist` has been called. 

I can't reproduce your problem. When I run `gulp serve:dist` I see the `app.js` file inside my `dist/scripts`.

**Comment by gazal-k** [comment]
Sorry, my bad. Hadn't updated npm packages. Perhaps an outdated version of gulp-useref was causing the issue
